### DISCUSSION TITLE
Make audio rate adjustments not affect timing windows

### DISCUSSION THREAD

**Discussion by smoogipoo** [ppy/osu#13187] [Ideas]
In osu-stable, applying DT/HT causes the timing windows to get shorter/longer as the audio rate is adjusted. This is the case for all rulesets except mania, which applies the inverse adjustment to the timing windows to ensure they are the same length at any audio rate.
In lazer, this is applied to all rulesets.

At some point, we should discuss (with the community) whether this is something that should be applied to all rulesets, no rulesets, or conditionally based on a ruleset's preference.

Related issues:
https://github.com/ppy/osu/issues/12242

**Comment by smoogipoo** [comment]
Discussion will be had at some point in the future, there's no plans to change this immediately.

**Comment by vernonlim** [comment]
Thanks for listening and reopening this discussion smoogipoo.

In the recent community meeting there was talk of roadblocks to getting players to try out lazer, and in my opinion the lack of timing window inversion for rates is the main one keeping mania players away. Arbitrary rate adjustments from 0.5-2.0x is the absolute biggest advantage lazer could have over stable for players, because of how much the mode relies on them (through generated map difficulties). The easiest way to demonstrate that is to look at the 4K MWC 2021 mappools - the majority of the maps are on rates. Being able to play maps on them without the mapper needing to submit generated difficulties would be a godsend, and I'm very sure a sizable portion of the mania community would check lazer out if news spread that this was a thing (through a twitter post demonstrating it, for example).

However, due to the timing windows changing, rate changes in lazer are currently in a non-usable state for mania. As seen from the identical difficulty settings of the mapper submitted rates, mania players expect the map to be sped up, not the game, with the timing windows kept the same. No other VSRG I know of changes timing windows with rates, including osu! stable mania DT/HT. While using 'Difficulty Adjust' to edit the OD so that after the rate change it matches the original is possible, it is definitely not ideal as players have to calculate the right value for every OD/rate combination and apply it manually. Changing timing windows is simply unheard of in the scene, and I can guarantee not a single experienced player feels comfortable with them or finds them useful (including myself and the creator of the issue this discussion links to).

I really want more people to follow the dev team's calls to try out lazer more, but not only is there not much of a reason to over stable other than skinning, DT/HT isn't even usable by player standards. Timing window inversion would not only fix that, but it would make the aforementioned mapper submitted rates be built into the game — a massive reason to check it out. 

**Could you please consider implementing timing window inversion for mania?**

> whether this is something that should be applied to all rulesets, no rulesets, or conditionally based on a ruleset's preference.

On a side note, my own view on this is that it should be conditional and tied to whether 'AR' changes, because it's the philosophy difference between whether rates speed up the map or the game. In almost all VSRGs scroll speed doesn't change between rates (including lazer mania), increasing visual density. Rates are, or end up viewed as, the speeding up (BPM increase) of the map — not the game — and in the end it's natural that timing windows also don't change. In comparison, games like osu! (standard/taiko/catch) have rates increase the AR, keeping the visual density the same, and tighter timing windows are expected along with them due to the understanding that it speeds up the game. If the AR changes, OD changes, and the opposite.

EDIT: 
Not sure about the implementation feasibility, but I think I have a user friendly solution to the issue. Have a checkbox in the DT/HT settings that controls whether AR/OD(/HP?) scales with rates (whether inversion is applied). It could be on by default for osu/taiko/catch and off by default for mania, and whenever score submission comes online the non-default state could be unranked.

**Reply by smoogipoo** [reply]
@peppy I think we probably should implement the timing window inversion for mania, so that DT/HT don't affect timing windows. Agree or disagree?

It's fairly easy to implement in the context of _only_ DT and HT, with my biggest concern otherwise being how to implement time-winding mods. It seems like a proper implementation requires hit windows have to become a global concept which the mod would constantly adjust in e.g. `Update`.

**Reply by peppy** [reply]
I think I still need some time to think about this one (specifically which rulesets/how we should be applying this).

If we want to make it opt-in for a ruleset, I'd argue that it should be supported at a `DrawableRuleset` or `HitWindows` level, rather than handled by the mod?

ie. 

```csharp
/// <summary>
/// Whether hit windows should scale alongside gameplay rate adjustments.
/// If <c>false</c> hit windows will always be the same duration in real-time, regardless of gameplay rate.
/// </summary>
protected virtual bool RateAdjustsAffectHitWindow => true;
```

and override to `false` in osu!mania only (if that's where we want to start).

**Reply by smoogipoo** [reply]
The hardest part in all of this is deciding where the calculations should take place. `HitWindows` is attached to `HitObject` (the model class) for better or worse (probably for the worse at this point) so it doesn't know about mods. Yet DHOs use `HitObject.HitWindows.CanBeHit(timeOffset)` whenever they need to do hit logic.

So... Should `HitWindows` have knowledge of mods?

The way I've thought of this is opposite - allowing mods to control `HitWindows`, but then the question becomes how to do stuff like time ramp mods because `HitWindows` aren't mutable right now.

**Reply by peppy** [reply]
Can we pass the current rate into the `CanBeHit` and handle the behaviour inside the subclassed `HitWindows`? That seems like a pretty sane approach.

**Reply by smoogipoo** [reply]
So then `DrawableHitObject` becomes knowledgeable of mods? Also it means that it requires explicit ruleset implementation rather than a flag which you override as per your suggestion.

**Reply by peppy** [reply]
Hmm, I'm going to need to go through the current path of message propagation to answer this properly (my hope would be that the rate could be pulled from `GameplayClock` rather than the mods, meaning that knowledge of mods is not required). Will have a look next week.

**Reply by bdach** [reply]
> my hope would be that the rate could be pulled from GameplayClock rather than the mods

Unfortunately (at least for 100% correctness) the rate is not sufficient when the track rate is changing mid-playback (see wind up/down) - in those cases, the duration of the hitwindow depends on the track rate throughout the entire hit window. The difference isn't bound to be much over a hitwindow as they are usually short, but still.

See [this comment](https://github.com/ppy/osu/issues/10283#issuecomment-771241482) wherein I already struggled against this with the spinner SPM counter. If the track rate changes mid-playback then dividing/multiplying by track rate in a given time instant no longer gives useful values.

(Also see https://github.com/ppy/osu/pull/14831, although I haven't reviewed that one for correctness yet.)

**Reply by smoogipoo** [reply]
We don't want to pull the rate from the `GameplayClock`. Hit windows should not change when the user is adjusting the playback speed of replays.

**Reply by peppy** [reply]
Keep in mind we do have 

https://github.com/ppy/osu/blob/6301111fa3c1d274c68766d3c1456ebd74d95138/osu.Game/Screens/Play/GameplayClock.cs#L40-L44

which is required and used in a few places.

**Comment by XeZey** [comment]
I'm not sure if this is more of a development thing that's just pretty hard to implement (in that case I'm not sure if there is any point in bringing this up now), or whether it's seen as something not that important.  Either way i felt that this is a pretty undervalued change. In the current state, rate changes on more difficult maps are hardly usable, especially on bigger rate adjustments. For mania and vsrg games (for somewhat experienced players specifically), rate changes are such a core part of the gameplay, that I'd imagine this is way more important to these players than most issues that currently exist within the gamemode. Sorry if there is no value in this comment, but I thought it's important to mention this since there hasn't been any activity on this as of recently and I don't know if there are any plans to change this, despite the fact that it seems like it would be a really big deal to most active players.

**Reply by peppy** [reply]
It's definitely still something on our radar. I'll link it into our current epic tracking breaking gameplay changes we want to review (https://github.com/ppy/osu/issues/18894)

**Comment by mk56-spn** [comment]
I've spent the last few weeks mulling over this topic and the sheer amount of issues that are created by speeding up / slowing down the game is astounding, and its going to be a MUCH bigger issue in Lazer than it ever had even the chance of being in stable. I'll detail what I mean:

using clock speed for rate mods will have massive implications with the amount of possible rates ( assuming that in ranked the allowed rates are each 0.05 change in speed that's  an increase in amount of possible rates from 3 to around 30 , in other words, there is 1000% the amount of rates that stable has, 500% if .1 increases end up being the selected amount) 30 rates * 110 AR states  = 3300 different animation speeds , some that vary a lot from each other in terms of feel ( 1.25 rate AR10 feels NOTHING like 1.9 rate AR10 )  and some that will barely be noticeable. from my experience I can say that DAs normal clock speed AR 10.5 and DT 1.5 x  (AR 10.5) are different enough to cause the player note-able  discomfort when they are trying to have the same experience on different maps but cant because of how DT is coded

next major issue: lazer currently has dozens of mods,, a  handful of which  (blinds , repel, magnetise, etc... ) have hard coded animations that do not account for rate changes, said animations may have been made to be pleasant at 1.0 speed. but it certainly holds up very poorly at 2x and .5x, which are extreme examples, but still, cases that every user will have access to.

Another big issue which ties in with the first is  that some OD and AR values will only be accessible at certain rates due to clock speed being used. for example. DA (AR 11)  with 1.0 rate is the maximum AR feel achievable at that song speed. now most people are fine with that because very few people play above AR 11. but what if you want to play AR 11 on a song at .5 speed, currently that is impossible

these issues can somewhat be tamed  with what mania does in stable by applying a counter effect, but it means going to every place the clock speed causes issue in any mod that is sensitive to it, and fixing it, which sounds like a major PITA.

I saw a comment by ppy ( which prompted this post coincidentally , but I cant seem to find it again)  explaining how DT could be changed to "squishing the map" ( bringing objects closer together in the timeline to simulate  the speed increase ) in order to avoid this issue (or well he i don't remember if he stated it, but its the logical deduction, since clock speed is no longer being changed) , which in all fairness he made rather clear he didn't want to/ had no intention at the time to do due to the overwhelmingly large code changes that it would require IIRC, not to mention server side stuff and other things I cant get my head around

my point in bringing this up is that i wanted to bring to light some of the issues  i through my playing experience have felt this to cause, so other people can perhaps weigh in on this, or if there is any other less drastic measures that can be taken that would void all these issues

finally i want to say that i do not wish to stir up controversy, i am well aware animation changes with DT are part of Osu's long standing culture and that i'm not putting them in a very good light in this post but i thought these were important considerations 

These are just my thoughts , that's all.

**Reply by smoogipoo** [reply]
Squishing the map is not the same as adding DT. It increases BPM. The former maintains the same density at different speeds, the latter maintains the same speed at different densities. Changing DT to affect BPM is not going to be anywhere near the same playing experience, and in general would actually ruin DT.

AR is completely different to both of those again, since it has a very specific domain upon which it operates - it changes the point in time where you start to see the hitobject (preempt). Nothing more. DT AR11 being slightly different to DA AR11 is intentional for this reason. DA mimics it correctly in terms of preempt alone, and tries its best to make the fade in appear good at AR>10, but not in terms of every other animation in the game; simply because it doesn't have control over that domain.

Keen eyes may also notice that DT AR<9 and DA AR<9 also feel different, it's the same "issue".

A separate mod that controls BPM would be a good idea - I implemented this in the past with a tool that a lot of osu! players used, and there's already been a few feature requests for it in lazer.

Whether DT should affect timing windows or not is, imo, an isolated matter. You could argue "yes" due it being a uniform change of speed, but you could also argue for "no" because uniformity doesn't matter a whole lot for it - it's a hidden variable.

**Comment by Zyfarok** [comment]
In osu! standard, I believe the OD increase with DT is perceived as normal and making sense. Indeed, for a player of a given rank, the maps that he can DT are usually lower star-rating (before DT) and thus lower OD too, and the OD increase from DT help bringing them closer to the OD you would play in nomod. Also, higher rank players are used to play HR which usually caps the OD at 10, and thus is comparable to the ODs they play in DT.

Thus I would expect the OD increase from DT to stay the DEFAULT behaviour in standard with DT/HT and all rate adjusts.
However this can lead to insanities when used with very high rate adjust (>1.5), thus something needs to be done.
Here are my suggestions:
- Cap the rate-adjusted OD such that a map can not have a smaller hit window than the current extremum in stable ("OD 11.11" or OD 10 + DT).
- Allow to disable the OD scaling of DT/HT/Rate-Adjust, and/or allow Difficulty-Adjust to be applied AFTER the rate change, thus overriding the OD. (both would solve the problem, but I don't know which one would be prefered)

And yes, I believe Wind-Up should have the same cap and maybe also allows the same override features as Rate-Adjust.

The same logic applies to AR. You can read this whole answer again by replacing OD (resp. "OD 11.11" and "hit window") with AR (resp. "AR 11" and "approach time") and everything applies (same reasoning, same problems, same solutions).

**Reply by Rekunan** [reply]
I disagree with this fundamentally.

One of my main reasons for looking forward to lazer is seeing those extreme AR's and OD's. Setting DA to AR11 and DT to 2.0x nets us AR12, which is something I really want to see players improve and perform at. Heck, I've seen people in mcosu go all the way to AR12.5. I think instead of forcing a cap, we should encourage the extension of the meta, because now that DA will allow people to play on AR11/OD11 without being locked to each other, the new max lock is requiring rate adjust for anything above 11.

**Reply by Zyfarok** [reply]
The cap is not strictly necessary so yeah if people don't want a cap we can leave it uncapped as people can fix that with difficulty adjust if they want. The mod multiplier might need to be adjusted so that increasing the speed while "capping" the AR doesn't lower the multiplier.

AFAIK, AR 12 is the circles starting to fade-in 150ms before the center of the hit-window, which is already above what is possible to humanly react to, thus at AR 12+ the player can only use the only thing that appears before the hit-circles: follow-points. This is quite ridiculous at this point which is why I think it makes more sense to cap to 11 by default, but yeah it might no be necessary.
Also, in any case, there's no reason to cap the AR in difficulty adjust to 11, that's for sure, and it would be fun to be able to go higher. (as far as 12.5 ?)

Same reasoning for OD, but I don't know if there's any known cap to humanly achievable UR.

**Reply by Rekunan** [reply]
Wow I completely glossed over your second suggestion, my bad.

I think I've seen that be suggested somewhere before, so there's at least some demand for that, would def be better than cap.

**Comment by smoogipoo** [comment]
Closing this discussion for now as the title of it and the mentioned issue has been addressed via https://github.com/ppy/osu/pull/24636 . Feel free to start another discussion if you want to reconsider this as a future lazer-only non-classic-mod gameplay change.
